dap/thread.go

```diff
@@ -217,6 +224,46 @@ func (t *thread) getStackFrame(dgst digest.Digest) *frame {
 	return f
 }
 
+func (t *thread) determineParent(op *pb.Op) int {
+	// Another section should have already checked this but
+	// double check here just in case we forget somewhere else.
+	// The rest of this method assumes there's at least one parent
+	// at index zero.
+	if len(op.Inputs) == 0 {
+		return -1
+	}
+
+	switch op := op.Op.(type) {
+	case *pb.Op_Exec:
+		for _, m := range op.Exec.Mounts {
+			if m.Dest == "/" {
+				return int(m.Input)
+			}
+		}
+		return -1
+	case *pb.Op_File:
+		// Determine possible candidate parents from the actions.
+		// All actions need to be consistent in which input is the
+		// direct parent.
+		candidates := make(map[int]struct{})
+		for _, action := range op.File.Actions {
+			candidates[int(action.Input)] = struct{}{}
+		}
+
+		if len(candidates) != 1 {
```

Comment:
Multiple actions are possible, even in Dockerfile. For example `COPY` with multiple source paths. Non-negative input of first action should be ok.

Reply:
The way it works is that input keys in action point first to the input array of the op and after that is exhausted then then index is result of another action https://github.com/moby/buildkit/blob/master/solver/pb/ops.proto#L310. So if you have two copy actions after each other it is smth like

```
fileop {
  inputs = [ A , B ]
  actions = [
     copy{input: 0, secondaryInput: 1},
     copy{input: 2, secondaryInput: 1}
```

where the 2 means the result of the first action `len(inputs)+0`

Reply:
I've updated this to just use the first suitable input from the list of file actions.